new CI infra with private ELK per test

AmitShaul opened the pull request:
## Motivation
Make the CI great again
merged -->

## Changes
added a new infra for running CI tests with private elk cluster, after test is finished all tests data is being sent to a central elk cluster.
in this repo I added another folder parallel to "tests", this is only for testing the new code.
in case of success "tests" will be overwritten.
makefile and ci.yml were changed to run both of our CIs , the new and the old, to run in parallel to test the new code without touching the old one.

THIS PR IS FOR TESTING the new code and not the final code!

AmitShaul committed:
WIP merged new CI code into tests_elk folder

AmitShaul committed:
removed deprecated extensions/v1beta1, replaced it with app/v1. WIP, changed ci.yml and Makefile to run new CI infra

AmitShaul committed:
fixed makefiles new elk pytest paths

AmitShaul committed:
added a secret for new cluster zone

AmitShaul committed:
added a new docker file for new infra, DockerFileTestsElk

lrettig commented:
Hi @AmitShaul, could you briefly summarize here the difference between the current testing infrastructure and this new ELK-based infrastructure? What's the difference/benefit for switching over? What problem(s) is it intended to address?

Regarding the PR, I'm a little confused about why you needed to duplicate the tests files, dockerfiles, etc. If these changes are intended to eventually replace the existing infrastructure, then we should update the existing files in place. If the intention is to run in parallel for a while, then we can just do that using a separate git branch, can't we?

BTW none of the new tests are currently being triggered by the CI here since it only runs the stage1 tests inside a PR. There are a few options if you want to try running the system tests. You could run the system tests locally, you could try [using `act`](https://github.com/spacemeshos/go-spacemesh/blob/develop/ci.md#testing-locally), or you could change the ci.yml workflow file here to trigger all of the tests in this branch. Let me know if you need a hand with this!

noamnelke commented:
@lrettig the idea was to run the new env in parallel for a while, but make it not interfere with normal tests. The way it's set up now, it should run in parallel, but bors won't depend on it (please comment if you don't think it will do this).

noamnelke commented:
bors try

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into dev_elk

AmitShaul committed:
fixed config file files paths, ready for testing in parallel with current CI

AmitShaul commented:
bors try

AmitShaul committed:
fixed a bug in filebeat files, NAMESPACE was switched with a specific namespace. fixed NAMESPACE search regex to negative look behind and search for _

AmitShaul commented:
bors try

AmitShaul committed:
handle exceptions in es dump

AmitShaul commented:
bors try

noamnelke commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

lrettig commented:
> @lrettig the idea was to run the new env in parallel for a while, but make it not interfere with normal tests. The way it's set up now, it should run in parallel, but bors won't depend on it (please comment if you don't think it will do this).

My concern here is around code duplication. I can't compare the old and new tests without a diff. Rather than duplicating all of the code, can't we parameterize the difference and move it into env vars or Makefile or a config file or something?

If that's not possible for some reason then my Plan B would be to keep this in a separate branch rather than duplicating the code.

AmitShaul committed:
changed k8s poet branch from develop to 73488d6

AmitShaul commented:
bors try

AmitShaul commented:
@lrettig this is a temporary setup to make both code run in parallel. 
I agree that code duplication is a horrible thing (it was really hard for me to get that going).
once we know that my code work I will open a new PR without the code duplication and reverse this PR.

AmitShaul committed:
DEBUGGING create dependencied between ci.yml so not all tests run in parallel, set timeout for ES dumping function

lrettig commented:
> I agree that code duplication is a horrible thing (it was really hard for me to get that going).
> once we know that my code work I will open a new PR without the code duplication and reverse this PR.

Okay, gotcha, thanks for clarifying. I know that it's sometimes easier to run these tests using CI, but as a reminder, you can definitely run the tests locally, either directly (e.g., `make dockertest-late-nodes`) or [using act](https://github.com/spacemeshos/go-spacemesh/blob/develop/ci.md#testing-locally).

I'll convert this PR into a draft now to make it clear that it isn't intended to be merged.

lrettig committed:
Merge branch 'develop' into dev_elk

lrettig committed:
Move elk tests to a new stage of its own

lrettig commented:
Per separate conversation with @AmitShaul I moved the new elk tests into their own "stage" in the CI which runs only after the existing system tests - hopefully this won't overload the system with too many tests running in parallel. Or maybe instead we want to _replace_ the existing tests? If these do not depend on them and this PR is only for testing, that would make sense - it would save resources and time.

AmitShaul committed:
changed new CI tests to work regardless to the success of the old CI tests

AmitShaul commented:
bors try

noamnelke commented:
@lrettig eventually, we'll replace the existing tests with the new ones. For now, we want to run them in parallel - to test the new tests.

Since they run on completely different infrastructure, there's no harm in running them in parallel, but there's also no strong incentive to do so - whatever is more convenient.

lrettig commented:
> Since they run on completely different infrastructure, there's no harm in running them in parallel, but there's also no strong incentive to do so - whatever is more convenient.

Hmm, @AmitShaul suggested that he _doesn't_ want them to all run in parallel since, when they do, tests seem to fail. So now they're running serially, as part of a new "phase" of tests, after the existing tests.

In any case, if the new tests don't actually depend upon the old tests in any way, for purposes of testing it may be faster/easier/cheaper to just _replace_ the old tests with the new ones.

AmitShaul committed:
WIP integrated fluent bit though with a bug - daemonset is pulling log msgs from all namespaces. added some sleep time before p2ps first test in order to let ES cluster to start. changed ci.yml to start 4 new infra tests simultaniously.

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into dev_elk

AmitShaul committed:
added lanes changes from f790a44e27ac7ef5cef82a72718971ac54416fd3 into tests_elk

AmitShaul committed:
WIP
added namespace labels to bootstrap client and poet k8s config files,
removed nodeSelector from fluentbit daemonset.
fixed fluentbit cluterrolebinding deletion.
changed NAMESPACE handling in deployment - instead of changing the original file, only duplicate files with the given pattern, replace the pattern and delete the file afterwards.

noamnelke commented:
>In any case, if the new tests don't actually depend upon the old tests in any way, for purposes of testing it may be faster/easier/cheaper to just replace the old tests with the new ones.

Amit already tested the tests and they worked. We wanted to merge this PR so that the tests run in addition to the old tests, so that we gain confidence that they pass when they should and fail when they should in the real world before replacing the old tests for good.

AmitShaul committed:
namespace support for fluent bit, now fluent bit will only collect logs from its own namespace.
kube deployment files with modified namespaces deletion happens right after file is being loaded to memory.
added tolerance to elastic logging k8s files.

AmitShaul committed:
added sleep in GitHub actions config yml file

AmitShaul commented:
bors try

AmitShaul committed:
changed timing on github actions file, fixed import bug in tests_elk wallet_api

AmitShaul commented:
bors try

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into dev_elk

AmitShaul committed:
test_p2p, bug fix, changed expected return value in send_msgs

AmitShaul committed:
This commit should be reverted, this is only for testing the new CI timing, commented up_pods assertion in analyse.py to avoid a known bug

AmitShaul commented:
bors try

AmitShaul committed:
[*] changed request and limit for CPU and memory to be equal in all of the full node tests
[*] changed statefulset dep to change NAMESPACE to the current run namespace
[*] removed genesis voting test, this test is depricated especially after Antons' changes when you don't make blocks for the first two epochs

AmitShaul committed:
added dummy test

AmitShaul commented:
bors try

AmitShaul committed:
[*] replaced Fluent-Bit daemonset podAffinity with nodeSelector
[*] wait for ES and Logstash to be ready, add this waiting time to GENESIS_TIME. now early logs arrive as expected.
[*] undo the commented assertion for nodes up in analysis.py

AmitShaul commented:
bors try

AmitShaul committed:
added an offset to wait_genesis, added timeout for elk waiting function

AmitShaul commented:
bors try

AmitShaul committed:
moved dummy test after stage 2, raised dummy test timeout to 5 minutes

AmitShaul commented:
bors try

AmitShaul committed:
fixed dummy test github actions file ci.yml, moved mining test to start instead of p2p test

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul committed:
fixed test dummy and set number of clients to 200, commented out es dumping function until ready

AmitShaul commented:
bors try

AmitShaul committed:
added sleep for debugging p2p test, this should be removed, added try except when waiting for ELK to be ready

AmitShaul commented:
bors try

AmitShaul committed:
test_many_gossip_dim, raised propagation sleep time to 50 instead of 20, removed sleep in the same function

AmitShaul commented:
bors try

AmitShaul committed:
raised test_many_gossip_sim propagation timeout to 240 instead of 40 (originally 20)

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul committed:
raised prop_sleep_time to 250, only adding this change so bors will try running CI with this code

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-

AmitShaul commented:
bors try-

AmitShaul commented:
bors try-

AmitShaul commented:
bors try-

AmitShaul commented:
bors try

AmitShaul committed:
raised sleep for propagation to 300, changed ci.yml flow

AmitShaul commented:
bors try

lrettig commented on .github/workflows/ci.yml:
```diff
@@ -209,3 +210,95 @@ jobs:
+  ## stage4: this is for testing the new CI env
+  systemtest-dummy:
+    # only run on push, not on pull_request
+    if: always() && github.event_name == 'push'
```
Remind me why the `always()` is necessary here. Do you want this to run even if stage1 or stage2 fails? If so, the `needs: ci-stage2` would be an issue.

lrettig commented on .github/workflows/ci.yml:
```diff
@@ -209,3 +210,95 @@ jobs:
+  systemtest-hare-elk:
+    # only run on push, not on pull_request
+    if: always() && github.event_name == 'push'
+    needs: systemtest-blocks-add-node-elk
```
Is there a reason it depends on this specific prior test? Or is this arbitrary?

lrettig commented on .github/workflows/ci.yml:
```diff
@@ -209,3 +210,95 @@ jobs:
+  systemtest-sync-elk:
+    # only run on push, not on pull_request
+    if: always() && github.event_name == 'push'
+    needs: systemtest-blocks-add-node-elk
```
Same question here as previous question.

lrettig reviewed (commented):
See questions!

AmitShaul committed:
DEBUG, this commit should be reverted, added 5000 secs sleep time in send_msgs for debugging

AmitShaul commented:
bors try

AmitShaul committed:
removed previous debugging commit, fixed fluent bit input file path only for sm files

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try

AmitShaul commented:
bors try-
